File: leetcode/0052-n-queens-ii.ts
/**
 * Solves the N-Queens problem and returns the number of distinct solutions
 * @param n - The size of the chessboard (n x n) and number of queens to place
 * @returns The total number of valid N-Queens solutions
 */
function totalNQueens(n: number): number {
    // Track which columns are occupied by queens
    const columnsOccupied: boolean[] = Array(10).fill(false);
  
    // Track which main diagonals (top-left to bottom-right) are occupied
    // Main diagonals have constant row + column sum
    const mainDiagonals: boolean[] = Array(20).fill(false);
  
    // Track which anti-diagonals (top-right to bottom-left) are occupied
    // Anti-diagonals have constant row - column difference
    const antiDiagonals: boolean[] = Array(20).fill(false);
  
    // Counter for valid solutions
    let solutionCount: number = 0;
  
    /**
     * Depth-first search to place queens row by row
     * @param currentRow - The current row index where we're trying to place a queen
     */
    const placeQueens = (currentRow: number): void => {
        // Base case: all queens successfully placed
        if (currentRow === n) {
            solutionCount++;
            return;
        }
      
        // Try placing a queen in each column of the current row
        for (let column: number = 0; column < n; column++) {
            // Calculate diagonal indices
            const mainDiagonalIndex: number = currentRow + column;
            const antiDiagonalIndex: number = currentRow - column + n; // Add n to ensure positive index
          
            // Check if current position conflicts with existing queens
            if (columnsOccupied[column] || 
                mainDiagonals[mainDiagonalIndex] || 
                antiDiagonals[antiDiagonalIndex]) {
                continue;
            }
          
            // Place queen at current position
            columnsOccupied[column] = true;
            mainDiagonals[mainDiagonalIndex] = true;
            antiDiagonals[antiDiagonalIndex] = true;
          
            // Recursively place queens in the next row
            placeQueens(currentRow + 1);
          
            // Backtrack: remove queen from current position
            columnsOccupied[column] = false;
            mainDiagonals[mainDiagonalIndex] = false;
            antiDiagonals[antiDiagonalIndex] = false;
        }
    };
  
    // Start placing queens from the first row
    placeQueens(0);
  
    return solutionCount;
}

File: leetcode/0052-n-queens-ii.java
class Solution {
    // Board size
    private int n;
    // Counter for valid solutions
    private int ans;
    // Track which columns are occupied by queens
    private boolean[] cols = new boolean[10];
    // Track which main diagonals (top-left to bottom-right) are occupied
    // For cells on the same main diagonal: row + col is constant
    private boolean[] dg = new boolean[20];
    // Track which anti-diagonals (top-right to bottom-left) are occupied
    // For cells on the same anti-diagonal: row - col is constant
    // Add n to make indices non-negative
    private boolean[] udg = new boolean[20];

    /**
     * Calculate the total number of distinct solutions to the n-queens puzzle
     * @param n The size of the chessboard (n x n)
     * @return The number of distinct solutions
     */
    public int totalNQueens(int n) {
        this.n = n;
        dfs(0);
        return ans;
    }

    /**
     * Depth-first search to place queens row by row
     * @param row Current row index being processed
     */
    private void dfs(int row) {
        // Base case: all queens successfully placed
        if (row == n) {
            ++ans;
            return;
        }
      
        // Try placing a queen in each column of the current row
        for (int col = 0; col < n; ++col) {
            // Calculate diagonal indices
            int mainDiagonal = row + col;
            int antiDiagonal = row - col + n;
          
            // Check if current position conflicts with existing queens
            if (cols[col] || dg[mainDiagonal] || udg[antiDiagonal]) {
                continue;
            }
          
            // Place queen at (row, col)
            cols[col] = true;
            dg[mainDiagonal] = true;
            udg[antiDiagonal] = true;
          
            // Recursively place queens in the next row
            dfs(row + 1);
          
            // Backtrack: remove queen from (row, col)
            cols[col] = false;
            dg[mainDiagonal] = false;
            udg[antiDiagonal] = false;
        }
    }
}